File: backend/app/services/notification_service.py
"""Envio de notificacao por Telegram e WhatsApp, com fallback entre canais.

O erro de cada canal e traduzido para uma frase acionavel antes de subir - a
mensagem original costuma trazer token ou URL, que nao podem aparecer na tela.
"""

import httpx
from html import escape
from loguru import logger
from typing import Optional, List
from ..models.schemas import NotifConfig, NotifResult, CalendarEvent


def _telegram_error_message(status_code: int, description: str) -> str:
    """Traduz erros conhecidos sem devolver token ou URL sensivel."""

    detail = description.strip()
    lowered = detail.lower()
    if status_code == 401 or "unauthorized" in lowered:
        return "Token do bot invalido. Gere ou copie novamente o token no BotFather."
    if "chat not found" in lowered:
        return (
            "Chat ID nao encontrado. Abra a conversa com o bot no Telegram, "
            "envie /start e confira o Chat ID."
        )
    if status_code == 403 or "bot was blocked" in lowered:
        return "O bot foi bloqueado ou nao tem permissao para enviar a esse chat."
    if status_code == 429 or "too many requests" in lowered:
        return (
            "O Telegram limitou os envios. Aguarde alguns instantes e "
            "tente novamente."
        )
    if "can't parse entities" in lowered:
        return "O Telegram recusou a formatacao da mensagem."
    if detail:
        return f"Telegram recusou o envio: {detail}"
    return f"Telegram recusou o envio (HTTP {status_code})."


def _telegram_text(
    message: str,
    config: NotifConfig,
    event: Optional[CalendarEvent],
    assistant_name: str,
) -> str:
    text = f"<b>{escape(assistant_name)}</b>\n\n{escape(message)}"
    if event and config.include_link and event.meeting_url:
        safe_url = escape(event.meeting_url, quote=True)
        text += f'\n\n🔗 <a href="{safe_url}">Entrar na reunião</a>'
    return text


async def _deliver_telegram(
    message: str,
    config: NotifConfig,
    event: Optional[CalendarEvent] = None,
    assistant_name: str = "Assistant",
) -> tuple[bool, str]:
    if not config.telegram_token or not config.telegram_chat_id:
        return False, "Preencha o token do bot e o Chat ID."

    try:
        async with httpx.AsyncClient(timeout=15) as client:
            resp = await client.post(
                f"https://api.telegram.org/bot{config.telegram_token}/sendMessage",
                json={
                    "chat_id": config.telegram_chat_id,
                    "text": _telegram_text(message, config, event, assistant_name),
                    "parse_mode": "HTML",
                    "disable_web_page_preview": False,
                },
            )
        try:
            data = resp.json()
        except ValueError:
            return False, f"Resposta invalida do Telegram (HTTP {resp.status_code})."
        if not data.get("ok"):
            return False, _telegram_error_message(
                resp.status_code,
                str(data.get("description", "")),
            )
        return True, "Telegram conectado e mensagem de teste enviada."
    except httpx.TimeoutException:
        return False, "O Telegram demorou para responder. Tente novamente."
    except httpx.HTTPError:
        return False, "Nao foi possivel conectar ao Telegram a partir do backend."
    except Exception:
        return False, "Falha inesperada ao enviar a mensagem pelo Telegram."


async def send_telegram(
    message: str,
    config: NotifConfig,
    event: Optional[CalendarEvent] = None,
    assistant_name: str = "Assistant",
) -> bool:
    """Envia a mensagem pelo bot do Telegram configurado."""
    ok, message_or_error = await _deliver_telegram(
        message,
        config,
        event,
        assistant_name,
    )
    if ok:
        logger.info(f"Telegram sent: {message[:60]}")
        return True
    logger.error("Telegram error: {}", message_or_error)
    return False


async def test_telegram_connection(config: NotifConfig) -> tuple[bool, str]:
    """Testa token e chat id do Telegram sem enviar lembrete de verdade.

    Returns:
        Uma tupla `(ok, mensagem)` para exibir direto na tela de configuracao.
    """
    return await _deliver_telegram(
        "✅ Assistente conectado! Notificações ativas.",
        config,
    )


async def send_whatsapp(message: str, config: NotifConfig) -> bool:
    """Envia a mensagem pelo provedor de WhatsApp configurado."""
    if not config.wa_number:
        return False
    try:
        match config.wa_provider:
            case "callmebot":
                return await _send_callmebot(message, config)
            case "zapi":
                return await _send_zapi(message, config)
            case "twilio":
                return await _send_twilio(message, config)
            case _:
                raise Exception(f"Provedor desconhecido: {config.wa_provider}")
    except Exception as e:
        logger.error(f"WhatsApp error: {e}")
        return False


async def _send_callmebot(message: str, config: NotifConfig) -> bool:
    number = config.wa_number.replace("+", "").replace(" ", "")
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(
            "https://api.callmebot.com/whatsapp.php",
            params={"phone": number, "text": message, "apikey": config.wa_token},
        )
    return resp.status_code == 200


async def _send_zapi(message: str, config: NotifConfig) -> bool:
    parts = config.wa_token.split(":")
    if len(parts) < 2:
        raise Exception("Z-API: formato token inválido (instance_id:token)")
    instance, token = parts[0], parts[1]
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.post(
            f"https://api.z-api.io/instances/{instance}/token/{token}/send-text",
            json={"phone": config.wa_number, "message": message},
        )
    return resp.status_code == 200


async def _send_twilio(message: str, config: NotifConfig) -> bool:
    import base64
    auth = base64.b64encode(f"{config.wa_sid}:{config.wa_token}".encode()).decode()
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.post(
            f"https://api.twilio.com/2010-04-01/Accounts/{config.wa_sid}/Messages.json",
            headers={"Authorization": f"Basic {auth}"},
            data={
                "From": "whatsapp:+14155238886",
                "To": f"whatsapp:{config.wa_number}",
                "Body": message,
            },
        )
    return resp.status_code == 201


async def send_notification(
    message: str,
    config: NotifConfig,
    channels: List[str] = ("telegram", "whatsapp"),
    event: Optional[CalendarEvent] = None,
    assistant_name: str = "Assistant",
) -> NotifResult:
    """Envia por todos os canais pedidos, respeitando a regra de fallback.

    Com `fallback_enabled`, o WhatsApp so e acionado quando o Telegram falhou; sem
    ele, os dois canais recebem. Falha de canal nao levanta excecao: volta dentro do
    `NotifResult`, para um canal quebrado nao impedir o outro.

    Args:
        message: texto ja formatado.
        config: preferencias de notificacao do usuario.
        channels: canais a tentar.
        event: evento de origem, quando o aviso vem da agenda.
        assistant_name: nome da persona, usado na assinatura da mensagem.

    Returns:
        O resultado por canal, com o erro de cada um quando houver.
    """
    result = NotifResult()

    if "telegram" in channels and config.telegram_enabled:
        try:
            result.telegram_ok = await send_telegram(message, config, event, assistant_name)
        except Exception as e:
            result.telegram_error = str(e)

    should_wa = (
        "whatsapp" in channels and config.wa_enabled
        and (not result.telegram_ok or not config.fallback_enabled)
    )
    if should_wa:
        try:
            result.whatsapp_ok = await send_whatsapp(message, config)
        except Exception as e:
            result.whatsapp_error = str(e)

    return result


def build_event_message(
    event: CalendarEvent,
    *,
    is_15min: bool,
    reminder_minutes: int = 15,
) -> str:
    """Escreve o texto do lembrete de um evento.

    Args:
        event: evento que gerou o aviso.
        is_15min: se e o aviso antecipado ou o do horario.
        reminder_minutes: antecedencia configurada, citada no texto.

    Returns:
        A mensagem pronta para envio.
    """
    time_str = event.start_time.strftime("%H:%M")
    src_map = {"google": "📗 Google Calendar", "teams": "📘 Teams", "outlook": "📙 Outlook"}
    src = src_map.get(event.source, event.source)
    if is_15min:
        return (
            f"⏰ Em {reminder_minutes} minutos:\n\n"
            f"📅 {event.title}\n🕐 {time_str}\n📌 {src}"
        )
    return f"🔔 Começando AGORA:\n\n📅 {event.title}\n🕐 {time_str}\n📌 {src}"

File: interface/lib/services/notification_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/app_config.dart';

class SendResult {
  final bool ok;

  /// Motivo da falha (ou canal usado no sucesso). Sem isso a notificação
  /// falhava em silêncio e não dava para saber o que corrigir.
  final String detail;

  const SendResult(this.ok, [this.detail = '']);

  String get summary => ok
      ? '✅ Notificação enviada${detail.isEmpty ? '' : ' via $detail'}!'
      : '❌ Não enviada — ${detail.isEmpty ? 'verifique as configurações' : detail}';
}

class NotificationService {
  final NotifConfig config;
  final String assistantName;

  NotificationService(this.config, this.assistantName);

  String buildEventMessage(CalendarEvent event, {required bool is15min}) {
    final when = is15min ? 'em ${config.reminderMinutes} minutos' : 'agora';
    final buf = StringBuffer();
    buf.write('🔔 [$assistantName] ${event.title} — $when');
    if (config.includeLink && event.meetingUrl != null) {
      buf.write('\n🔗 ${event.meetingUrl}');
    }
    return buf.toString();
  }

  Future<SendResult> send(String message, {CalendarEvent? event}) async {
    final problems = <String>[];

    if (config.tgEnabled) {
      if (config.tgToken.trim().isEmpty || config.tgChatId.trim().isEmpty) {
        problems.add('Telegram sem token ou Chat ID em Configuracoes');
      } else {
        final error = await _sendTelegram(message);
        if (error == null) return const SendResult(true, 'Telegram');
        problems.add('Telegram: $error');
      }
    }

    final canUseWhatsApp =
        config.waEnabled && config.waNumber.trim().isNotEmpty;
    if (canUseWhatsApp && (problems.isEmpty || config.fallbackEnabled)) {
      if (await _sendWhatsApp(message)) {
        return const SendResult(true, 'WhatsApp');
      }
      problems.add('WhatsApp recusou o envio');
    }

    if (problems.isEmpty) {
      return const SendResult(
        false,
        'nenhum canal ativo (ative Telegram ou WhatsApp em Configuracoes)',
      );
    }
    return SendResult(false, problems.join(' | '));
  }

  Future<bool> testTelegram() async {
    try {
      if (config.tgToken.trim().isEmpty || config.tgChatId.trim().isEmpty) {
        return false;
      }
      final error =
          await _sendTelegram('Assistente conectado! Notificacoes ativas.')
              .timeout(const Duration(seconds: 15));
      return error == null;
    } catch (_) {
      return false;
    }
  }

  /// Retorna null no sucesso ou uma mensagem explicando a recusa.
  Future<String?> _sendTelegram(String message) async {
    try {
      final url = Uri.parse(
        'https://api.telegram.org/bot${config.tgToken.trim()}/sendMessage',
      );
      final r = await http.post(url, body: {
        'chat_id': config.tgChatId.trim(),
        'text': message,
      }).timeout(const Duration(seconds: 10));
      final data = jsonDecode(r.body) as Map<String, dynamic>;
      if (data['ok'] == true) return null;
      return _telegramError(
        r.statusCode,
        data['description']?.toString() ?? '',
      );
    } on TimeoutException {
      return 'o Telegram demorou para responder';
    } catch (e) {
      return 'falha de conexao ($e)';
    }
  }

  /// Mesmas traducoes usadas pelo backend, sem expor token nem URL.
  static String _telegramError(int statusCode, String description) {
    final lowered = description.toLowerCase();
    if (statusCode == 401 || lowered.contains('unauthorized')) {
      return 'token do bot invalido (gere outro no BotFather)';
    }
    if (lowered.contains('chat not found')) {
      return 'Chat ID nao encontrado (envie /start para o bot e confira o ID)';
    }
    if (statusCode == 403 || lowered.contains('bot was blocked')) {
      return 'o bot foi bloqueado ou nao pode enviar nesse chat';
    }
    if (statusCode == 429 || lowered.contains('too many requests')) {
      return 'o Telegram limitou os envios; tente em instantes';
    }
    return description.isEmpty
        ? 'recusado pelo Telegram (HTTP $statusCode)'
        : 'recusado pelo Telegram ($description)';
  }

  String buildReminderMessage(CalendarEvent event, int minutesBefore) {
    final when = minutesBefore >= 60
        ? 'em ${minutesBefore ~/ 60}h${(minutesBefore % 60) == 0 ? '' : '${minutesBefore % 60}min'}'
        : 'em $minutesBefore minutos';
    final buf = StringBuffer('🔔 [$assistantName] ${event.title} — $when');
    if (config.includeLink && (event.meetingUrl?.isNotEmpty ?? false)) {
      buf.write('\n🔗 ${event.meetingUrl}');
    }
    return buf.toString();
  }

  Future<bool> _sendWhatsApp(String message) async {
    try {
      switch (config.waProvider) {
        case 'callmebot':
          final url = Uri.parse(
            'https://api.callmebot.com/whatsapp.php'
            '?phone=${Uri.encodeComponent(config.waNumber)}'
            '&text=${Uri.encodeComponent(message)}'
            '&apikey=${Uri.encodeComponent(config.waToken)}',
          );
          final r = await http.get(url).timeout(const Duration(seconds: 10));
          return r.statusCode == 200;
        default:
          return false;
      }
    } catch (_) {
      return false;
    }
  }
}

/// Lembrete agendado pelo botão Notificar do painel de próximos eventos.
class PendingEventReminder {
  final String eventId;
  final String title;
  final int minutesBefore;
  final DateTime fireAt;

  const PendingEventReminder({
    required this.eventId,
    required this.title,
    required this.minutesBefore,
    required this.fireAt,
  });
}

/// Agenda lembretes por evento com antecedência escolhida pelo usuário.
/// Os timers vivem enquanto o app estiver aberto (mesma limitação dos
/// lembretes automáticos da agenda) e cada evento tem no máximo um lembrete
/// manual: reagendar substitui o anterior.
class EventReminderScheduler {
  static final Map<String, Timer> _timers = {};
  static final Map<String, PendingEventReminder> _pending = {};

  static PendingEventReminder? pendingFor(String eventId) => _pending[eventId];

  /// Agenda (ou reagenda) o envio para [minutesBefore] minutos antes do
  /// início do evento e retorna o horário de disparo. Se esse horário já
  /// passou, o envio acontece imediatamente.
  static DateTime schedule({
    required CalendarEvent event,
    required int minutesBefore,
    required NotifConfig notif,
    required String assistantName,
  }) {
    cancel(event.id);
    final fireAt = event.startTime.subtract(Duration(minutes: minutesBefore));
    var delay = fireAt.difference(DateTime.now());
    if (delay.isNegative) delay = Duration.zero;

    _pending[event.id] = PendingEventReminder(
      eventId: event.id,
      title: event.title,
      minutesBefore: minutesBefore,
      fireAt: fireAt,
    );
    _timers[event.id] = Timer(delay, () {
      _timers.remove(event.id);
      _pending.remove(event.id);
      final service = NotificationService(notif, assistantName);
      unawaited(
        service.send(
          service.buildReminderMessage(event, minutesBefore),
          event: event,
        ),
      );
    });
    return fireAt;
  }

  static bool cancel(String eventId) {
    _timers.remove(eventId)?.cancel();
    return _pending.remove(eventId) != null;
  }
}
